cocotb/share/include/gpi_logging.h

```diff
@@ -53,10 +53,6 @@ enum gpi_log_levels {
 #define LOG_INFO(...)      gpi_log("cocotb.gpi", GPIInfo,          __FILE__, __func__, __LINE__, __VA_ARGS__);
 #define LOG_WARN(...)      gpi_log("cocotb.gpi", GPIWarning,       __FILE__, __func__, __LINE__, __VA_ARGS__);
 #define LOG_ERROR(...)     gpi_log("cocotb.gpi", GPIError,         __FILE__, __func__, __LINE__, __VA_ARGS__);
-#define LOG_CRITICAL(...)  do { \
-    gpi_log("cocotb.gpi", GPICritical,      __FILE__, __func__, __LINE__, __VA_ARGS__); \
-    exit(1); \
```

Comment:
Are we sure we don't just want to remove this line?

Reply:
No. Critical-level usually do terminate the program. It's not a good idea, but it is a widespread one. I'd rather just not have it there to reduce confusion.

* DEBUG -> verbose information useful to the developer about the program while it runs
* INFO -> information useful to the end user about the program while it runs
* WARN -> when you encounter a situation that may result in undefined or surprising behavior, or an error that you could immediately handle
* return code/exception -> when you encounter an error you can handle
* ERROR -> when you encounter an error you cannot handle
* CRITICAL -> ERROR, but contains some default way to end the program

CRITICAL assumes there is a simple parameter-less function that could be called to terminate the program (otherwise is hard to implement). In many implementations this ignores clean shutdown which is usually not a good idea (like Go). It makes ERROR moot since if you are using ERROR you have encountered an error you cannot handle, you probably should be trying to shut down.

Reply:
I'd argue that since:
* This is really just an interface to `logging.log` in python
* `logging.critical` does not terminate the program

That it is fine to use `critical` here.

> Critical-level logs usually do terminate the program.

I'd phrase this the other way - critical-level logs usually _immediately precede_ termination of the program. I wouldn't expect the logger to handle the termination itself.

Reply:
In several logging frameworks critical logs do terminate the program, I mentioned Go (there are others, just don't remember off the top of my head). And you are assuming that a Python logger will be the only implementation, the GPI should not tie itself to a particular language. In fact I have a change on a branch that implements a more black-boxed approach to the GPI logger and moves all the Python bits into the simulator module. I'm trying to move the GPI into a state where it could be used standalone, service multiple user, or be wrapped in another language.

You mention in certain places logs should be critical because we immediately shutdown, but I think you are missing the point. All `LOG_ERROR`s should be immediately followed by a shutdown. Where `LOG_ERROR` is used currently in user code (that doesn't immediately shut down), that is temporary until we decide on how to turn them into exceptions or error codes. There should never be a `LOG_ERROR` that isn't immediately followed by a shut down because `LOG_ERROR` means an error was detect that can't be handled; the program should not continue. In that case ERROR and CRITICAL fill them same role and one is redundant.

If we keep `LOG_CRITICAL`, `LOG_ERROR` will (should) disappear from the code. I don't really care what we call it, but we only need one.

Reply:
> Where LOG_ERROR is used currently in user code (that doesn't immediately shut down), that is temporary until we decide on how to turn them into exceptions or error codes. 

But we use `LOG_ERROR` from the python side to mean "a test failed". So if you insist on your model, then it sounds like `LOG_ERROR` should map to `log.critical` in python?

Reply:
`LOG_CRITICAL` would probably be the one to keep since it better parallels the intent. It might be a better idea to use `_sim_event` rather than log in error cases much like an exception. I think I need to think this through a bit more before confronting the issue.

I'll just follow your recommendation and make sure we are using `LOG_CRITICAL` in the proper places and do clean shutdown in those cases.